mesonbuild/backend/backends.py

```diff
@@ -1141,9 +1141,21 @@ def create_test_serialisation(self, tests: T.List['Test']) -> T.List[TestSeriali
                     cmd_args.extend(self.construct_target_rel_paths(a, t.workdir))
                 else:
                     raise MesonException('Bad object in test command.')
+
+            t_env = t.env
+            if not machine.is_windows() and not machine.is_cygwin() and not machine.is_darwin():
+                ld_lib_path: T.Set[str] = set()
+                for d in depends:
+                    if isinstance(d, build.BuildTarget):
```

Comment:
I'm not sure I'm asking you to fix `CustomTarget`, but I think removing the `isinstance` check here is appropriate.

Reply:
> I think removing the isinstance check here is appropriate

Hmm OK, I'm not sure if all `Target`s implement that though? Are there others than Build+Custom?

> Also, don't you need to add a an entry to ld_lib_path for d?

Not AFAICT. If it's a `SharedLibrary` then it already adds itself before summing up the children.

Reply:
WEll, I don't think `Target` is actually the correct annotation. `depends` can be `BuildTarget | CustomTarget | CustomTargetIndex | GeneratedList` (if my reverse engineering of the types are correct). `GeneratedList` isn't a `Target` at all, so I think you could make the check `if isinstance(d, build.Target)` and that would work. The only other type is a `RunTarget`, but you can't depend on those.

Reply:
Computer says no, I'm afraid:
```
~/src/misc/meson test-smash-ld-library-path * % ./run_mypy.py
Running mypy (this can take some time) ...
mesonbuild/backend/backends.py:1151:34: error: "Target" has no attribute "get_all_link_deps"  [attr-defined]
```

Should I keep what I have, or add another commit which adds a no-op `get_all_link_deps` to the `Target` base?

Reply:
... I hate our inheritance model. I have some very length patches to fix that, but I guess for now you'll have to leave the isinstance check. Sigh.

Reply:
You can do an assert instead.

Reply:
I can't:
```
~/src/misc/meson test-smash-ld-library-path * % PAGER= git diff
diff --git a/mesonbuild/backend/backends.py b/mesonbuild/backend/backends.py
index 747d80edd..709a36a12 100644
--- a/mesonbuild/backend/backends.py
+++ b/mesonbuild/backend/backends.py
@@ -1147,7 +1147,8 @@ class Backend:
             if not machine.is_windows() and not machine.is_cygwin() and not machine.is_darwin():
                 ld_lib_path: T.Set[str] = set()
                 for d in depends:
-                    if isinstance(d, build.BuildTarget):
+                    if isinstance(d, build.Target):
+                        assert(any([isinstance(d, build.BuildTarget), isinstance(d, build.CustomTarget), isinstance(d, build.CustomTargetIndex)]))
                         for l in d.get_all_link_deps():
                             if isinstance(l, build.SharedLibrary):
                                 ld_lib_path.add(os.path.join(self.environment.get_build_dir(), l.get_subdir()))
~/src/misc/meson test-smash-ld-library-path * % ./run_mypy.py
Running mypy (this can take some time) ...
mesonbuild/backend/backends.py:1152:34: error: "Target" has no attribute "get_all_link_deps"  [attr-defined]
```

Reply:
Hm OK, changing the `any` to a set of `or`s helps. I don't mind doing either; the assert looks ugly to me, but it's your sense of taste as reviewers that counts rather than mine, so let me know which you prefer.

Reply:
I haven't looked closely at what the code is trying to do. But if the `if` itself is just there to help mypy typecheck, then assertions are usually better. If you have to actually do branching, then `if` is better.

Are there valid arguments that could end up in this variable and don't have the get_all_link_deps method? Why is GeneratedList relevant here anyway -- I didn't think tests accepted that.

Reply:
The `if` is required for branching, yeah. The `assert` is purely for making mypy not scream. I'm not sure there's anything I can change at this point?

Reply:
isninstace can take a tuple of types to check: `isinstance(d, (build.BuildTarget, build.CustomTarget, build.CustomTargetIndex))`, JFYI

Reply:
Ah thankyou! 👨🏻‍🎓